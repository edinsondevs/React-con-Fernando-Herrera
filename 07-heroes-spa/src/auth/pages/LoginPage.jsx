import React, { useContext } from "react";
import { useNavigate } from "react-router-dom";
import { AuthContext } from "../context/AuthContext";

export const LoginPage = () => {
	const { login } = useContext(AuthContext);
	const navigate = useNavigate();
	
	const onLogin = () => {
		login('Edinson Madrid');
		navigate("/", {
			replace: true,
		});
	};

	return (
		<div className='container mt-5'>
			<h1>LoginPage</h1>
			<hr />
			<button
				type='button'
				className='btn btn-outline-primary'
				onClick={onLogin}>
				Login
			</button>
		</div>
	);
};
