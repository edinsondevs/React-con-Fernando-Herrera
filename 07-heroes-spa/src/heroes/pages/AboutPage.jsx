import React from 'react'

export const AboutPage = () => {
  return (
    <h1 data-testid='about-page'>AboutPage</h1>
  )
}
