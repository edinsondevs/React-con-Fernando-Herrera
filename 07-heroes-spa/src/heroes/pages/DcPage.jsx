import React from 'react'
import { HeroList } from '../components';

export const DcPage = () => {
  return (
		<>
			<h1 data-testid='dc-comics'>Dc Comics</h1>
			<hr />
			<HeroList publisher='DC Comics' />
		</>
  );
}
