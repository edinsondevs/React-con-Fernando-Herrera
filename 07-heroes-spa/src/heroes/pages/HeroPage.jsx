import { useNavigate, useParams } from "react-router-dom";
import { getHeroById } from "../helpers";
import { useMemo } from "react";

export const HeroPage = () => {
	const navigate = useNavigate();
	
	const { id } = useParams();
	
	const hero = useMemo( () => getHeroById(id), [ id ]);
		
	const onNavigateBack = () => {
		navigate(-1)
	};
	
	if (!hero) {
		<h2 className='container d-flex justify-content-center'>
			404 - Not Found
		</h2>
		
	}
	return (
		<div className='row mt-5 animate__animated animate__fadeInLeft'>
			<div className='col-4'>
				<img
					alt=''
					src={`/assets/heroes/${id}.jpg`}
					className='img-thumbnail'
				/>
			</div>
			<div className='col-8'>
				<h3>{hero.superhero}</h3>
				<ul className='list-group list-group-flush'>
					<li className='list-group-item'>
						<b>Alter ego: </b> {hero.alter_ego}
					</li>
					<li className='list-group-item'>
						<b>Publisher: </b> {hero.publisher}
					</li>
					<li className='list-group-item'>
						<b>First Appearance: </b> {hero.first_appearance}
					</li>
				</ul>

				<h5 className='mt-3'>Characters: </h5>
				<p>{hero.characters}</p>

				<button
					className='btn btn-outline-dark'
					onClick={onNavigateBack}>
					Regresar
				</button>
			</div>
		</div>
	);
};
