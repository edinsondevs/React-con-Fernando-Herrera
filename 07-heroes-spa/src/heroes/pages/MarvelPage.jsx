import React from 'react'
import { HeroList } from '../components'

export const MarvelPage = () => {
  return (
    <>
      <h1 aria-label='marvel'>Marvel</h1>
      <hr />
      <HeroList publisher='Marvel Comics' />
    </>
  )
}
