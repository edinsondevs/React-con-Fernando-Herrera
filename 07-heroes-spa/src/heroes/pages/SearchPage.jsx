import { HeroCard } from "../components/HeroCard";
import { useNavigate } from "react-router-dom";
import { useForm } from "../../hooks/useForm/useForm";
import queryString from "query-string";
import { getHeroesByName } from "../helpers/getHeroesByName";

export const SearchPage = () => {
	const navigate = useNavigate();

	const { q = "" } = queryString.parse(location.search);
	const heroes = getHeroesByName(q);

	const { searchText, onInputChange, onResetForm } = useForm({
		searchText: "",
	});

	const onSearchSubmit = (event) => {
		event.preventDefault();
		// if (searchText.trim().length <= 1) return;

		navigate(`?q=${searchText}&asc=true`);

		onResetForm();
	};

	return (
		<>
			<h1>Search</h1>
			<hr />
			<div className='row'>
				<div className='col-5'>
					<h4>Searching</h4>
					<hr />
					<form onSubmit={onSearchSubmit}>
						<input
							type='text'
							name='searchText'
							placeholder='Search a hero'
							className='form-control'
							autoComplete='off'
							value={searchText}
							onChange={onInputChange}
						/>
						<button
							className='btn btn-outline-dark mt-4'
							onClick={() => {}}>
							Search
						</button>
					</form>
				</div>

				<div className='col-7'>
					<h4>Results</h4>
					<hr />
					{!q && (
						<div className='alert alert-primary'> Search Hero</div>
					)}
					{heroes.length === 0 && q && (
						<div className='alert alert-danger'>
							{" "}
							No Hero with <b> {q} </b>{" "}
						</div>
					)}
					{heroes.map((hero) => (
						<HeroCard
							key={hero.id}
							{...hero}
						/>
					))}
				</div>
			</div>
		</>
	);
};
