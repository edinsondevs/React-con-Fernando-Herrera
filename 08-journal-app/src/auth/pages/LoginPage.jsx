import Grid from "@mui/material/Grid";
import Typography from "@mui/material/Typography";
import TextField from "@mui/material/TextField";
import Button from "@mui/material/Button";
import Link from "@mui/material/Link";
import GoogleIcon from "@mui/icons-material/Google";
import { Link as RouterLink } from "react-router-dom";
import { AuthLayout } from "../layout/AuthLayout";

export const LoginPage = () => {
	return (
		<AuthLayout title='Login'>
			<form action=''>
				<Grid container>
					<Grid
						item
						xs={12}
						md={12}
						sx={{ mt: 2 }}>
						<TextField
							type='email'
							placeholder='correo@gmail.com'
							fullWidth
							label='correo'
							// value={}
							// onChange={}
						/>
					</Grid>
					<Grid
						item
						xs={12}
						md={12}
						sx={{ mt: 2 }}>
						<TextField
							type='password'
							placeholder='contraseña'
							fullWidth
							label='contraseña'
							// value={}
							// onChange={}
						/>
					</Grid>
					<Grid
						container
						spacing={2}
						sx={{ mt: 1, mb: 2 }}>
						<Grid
							item
							xs={12}
							sm={12}
							md={6}>
							<Button
								variant='contained'
								fullWidth>
								Login
							</Button>
						</Grid>
						<Grid
							item
							xs={12}
							sm={12}
							md={6}>
							<Button
								startIcon={<GoogleIcon />}
								variant='contained'
								fullWidth>
								<Typography variant='body1'>Google</Typography>
							</Button>
						</Grid>
					</Grid>
				</Grid>

				<Grid
					container
					direction='row'
					justifyContent='end'>
					{" "}
					<Link
						color='inherit'
						component={RouterLink}
						to='/auth/register'>
						Crear una cuenta
					</Link>
				</Grid>
			</form>
		</AuthLayout>
	);
};
