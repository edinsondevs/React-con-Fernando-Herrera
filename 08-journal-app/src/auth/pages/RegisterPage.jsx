import Grid from "@mui/material/Grid";
import Typography from "@mui/material/Typography";
import TextField from "@mui/material/TextField";
import Button from "@mui/material/Button";
import Link from "@mui/material/Link";
import { Link as RouterLink } from "react-router-dom";
import { AuthLayout } from "../layout/AuthLayout";
import { useForm } from "../../hooks";
import { useState } from "react";
import { startCreatingUserWithEmailPassword } from './../../store/auth/thunks';
import { useDispatch } from 'react-redux';
import { useSelector } from 'react-redux';
import { useMemo } from "react";
import { Alert } from "@mui/material";

const initialDataForm = {
	displayName: "",
	email: "",
	password: "",
};

// Validacion para el useForm del formulario
const formValidations = {
	displayName: [(value) => value.length >= 3, "Debe introducir un nombre valido" ],
	email: [ (value) => value.includes("@"), "Debe introducir un correo valido" ],
	password: [ (value) => value.length >= 6, "La contrase単a debe tener mas de 6 caracteres" ],
};

export const RegisterPage = () => {
	const [formSubmitted, setFormSubmitted] = useState(false);
	const { status, errorMessage } = useSelector(state => state.auth);
	const isCheckingAuthentication = useMemo(()=> status === 'checking',[status])
	const {
		formState,
		onInputChange,
		displayName,
		nameValid,
		email,
		emailValid,
		password,
		passwordValid,
		isFormValid,
	} = useForm(initialDataForm, formValidations);
	const dispatch = useDispatch();

	const onSubmit = (event) => {
		event.preventDefault();
		setFormSubmitted(true);
		if (!isFormValid) return;
		dispatch(startCreatingUserWithEmailPassword(formState));
	};
	
	return (
		<AuthLayout title='Crear cuenta'>
			<form onSubmit={onSubmit}>
				<Grid container>
					<Grid
						item
						xs={12}
						md={12}
						sx={{ mt: 2 }}>
						<TextField
							type='text'
							placeholder='Nombre completo'
							fullWidth
							label='Nombre completo'
							name='displayName'
							value={displayName}
							onChange={onInputChange}
							helperText={nameValid}
							error={nameValid && formSubmitted}
						/>
					</Grid>
					<Grid
						item
						xs={12}
						md={12}
						sx={{ mt: 2 }}>
						<TextField
							type='email'
							placeholder='correo@gmail.com'
							fullWidth
							label='Correo'
							value={email}
							name='email'
							helperText={emailValid}
							error={emailValid && formSubmitted}
							onChange={onInputChange}
						/>
					</Grid>
					<Grid
						item
						xs={12}
						md={12}
						sx={{ mt: 2 }}>
						<TextField
							type='password'
							placeholder='Contrase単a'
							fullWidth
							label='Contrase単a'
							value={password}
							name='password'
							onChange={onInputChange}
							helperText={passwordValid}
							error={passwordValid && formSubmitted}
						/>
					</Grid>
					<Grid
						container
						sx={{ mt: 1, mb: 2 }}>
					<Grid 
						item
						xs={12}
						display={ (!!errorMessage ) ? '' : 'none'}
					>
						<Alert severity="error" >
							{errorMessage}
						</Alert>
					</Grid>
						<Button
							variant='contained'
							type='submit'
							disabled={isCheckingAuthentication}
							fullWidth>
							Crear Cuenta
						</Button>
					</Grid>
				</Grid>

				<Grid
					container
					direction='row'
					justifyContent='end'>
					<Typography
						variant='subtitle2'
						sx={{ mr: 1 }}>
						多Ya tienes cuenta?
					</Typography>
					<Link
						color='inherit'
						component={RouterLink}
						to='/auth/login'>
						Ingresar
					</Link>
				</Grid>
			</form>
		</AuthLayout>
	);
};
