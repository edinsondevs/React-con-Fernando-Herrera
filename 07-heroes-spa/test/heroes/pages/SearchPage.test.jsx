import { screen, render, fireEvent } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { SearchPage } from "../../../src/heroes/pages/SearchPage";

// * MOCK DEL USENAVIGATE
const mockedUseNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
	...jest.requireActual("react-router-dom"),
	useNavigate: () => mockedUseNavigate,
}));

describe("Pruebas en SearchPage", () => {

    beforeEach(() => jest.clearAllMocks());

	test("debe de mostrarse correctamente con valores por defecto", () => {
		const { container } = render(
			<MemoryRouter>
				<SearchPage />
			</MemoryRouter>
		);
		expect(container).toMatchSnapshot();
        const label = screen.getByLabelText("alert-search")
        expect(label).toBeTruthy();

	});

	test("debe de mostrar batman y el input con el valor del queryString", () => {
		
        render(
			<MemoryRouter initialEntries={["/search?q=batman"]}>
				<SearchPage />
			</MemoryRouter>
		);
        expect( screen.getByText('Batman'))
        const img = screen.getByRole('img');
        expect(img.src).toContain("/assets/heroes/dc-batman.jpg");
        
        expect(screen.queryByLabelText("alert-search")).not.toBeTruthy();
	});

	test("debe de mostrar un error si no se encuentra el heroe", () => {
		
        render(
			<MemoryRouter initialEntries={["/search?q=batmandeprueba"]}>
				<SearchPage />
			</MemoryRouter>
		);
        expect(screen.getByLabelText("alert-danger")).toBeTruthy(); 
        
	});
    
	test("debe de llamar el navigate a la pantalla nueva", () => {
        const optionSearch = "superman";
        render(
            <MemoryRouter initialEntries={["/search"]}>
				<SearchPage />
			</MemoryRouter>
		);
        const input = screen.getByRole('textbox')
        fireEvent.change(input, { target: { value: optionSearch } });
        const btnSearch = screen.getByRole('button', {name: 'Search'});
        fireEvent.click(btnSearch);
        
        // const form = screen.getByLabelText('form');        
        // fireEvent.submit(form)
        
        expect(mockedUseNavigate).toHaveBeenCalled();
        expect(mockedUseNavigate).toHaveBeenCalledWith(`?q=${optionSearch}`);

	});
});
