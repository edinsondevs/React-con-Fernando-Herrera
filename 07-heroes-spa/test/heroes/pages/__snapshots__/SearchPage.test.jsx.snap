// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en SearchPage debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search a hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-dark mt-4"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        aria-label="alert-search"
        class="alert alert-primary"
      >
         Search Hero
      </div>
    </div>
  </div>
</div>
`;
